docs/src/modules/ROOT/pages/enterprise-edition/enterprise-edition.adoc

```diff
@@ -716,5 +717,215 @@ the host is likely to hang or freeze,
+                      .penalize(SimpleScore.ONE)
+                      .asConstraint("b");
+    }
+}
+----
+
+is transformed to a class that looks like this:
+
+[source,java,options="nowrap"]
+----
+public class MyConstraintProvider implements ConstraintProvider {
+    private static final Predicate<Shift> $predicate1 = shift -> shift.getEmployee().getName().equals("Ann");
+
+    public Constraint[] defineConstraints(ConstraintFactory constraintFactory) {
+        return new Constraint[] {
+            a(constraintFactory),
+            b(constraintFactory)
+        };
+    }
+
+    Constraint a(ConstraintFactory constraintFactory) {
+        return factory.forEach(Shift.class)
+                      .filter($predicate1)
+                      .penalize(SimpleScore.ONE)
+                      .asConstraint("a");
+    }
+
+    Constraint b(ConstraintFactory constraintFactory) {
+        return factory.forEach(Shift.class)
+                      .filter($predicate1)
+                      .penalize(SimpleScore.ONE)
+                      .asConstraint("b");
+    }
+}
+----
+
+IMPORTANT: This transformation prevents breakpoints from being placed inside the ConstraintProvider.
+
+This allows building blocks to share the same nodes without needing the `ConstraintProvider` to be written in an awkward way.
+
+==== Configuration
+
+[tabs]
+======
+Plain Java::
+* Add `ai.timefold.solver.enterprise:timefold-solver-enterprise-asm` to your `pom.xml`:
++
+[source,xml,options="nowrap"]
+----
+<!-- ... -->
+<dependencies>
+  <dependency>
+    <groupId>ai.timefold.solver.enterprise</groupId>
+    <artifactId>timefold-solver-enterprise-asm</artifactId>
+  </dependency>
+</dependencies>
+<!-- ... -->
+----
++
+* Add `<constraintStreamAutomaticNodeSharing>true</constraintStreamAutomaticNodeSharing>` in your `solverConfig.xml`:
++
+[source,xml,options="nowrap"]
+----
+<!-- ... -->
+<scoreDirectorFactory>
+  <constraintProviderClass>org.acme.MyConstraintProvider</constraintProviderClass>
+  <constraintStreamAutomaticNodeSharing>true</constraintStreamAutomaticNodeSharing>
+</scoreDirectorFactory>
+<!-- ... -->
+----
+
+Spring Boot::
+* Add `ai.timefold.solver.enterprise:timefold-solver-enterprise-asm` to your `pom.xml`:
++
+[source,xml,options="nowrap"]
+----
+<!-- ... -->
+<dependencies>
+  <dependency>
+    <groupId>ai.timefold.solver.enterprise</groupId>
+    <artifactId>timefold-solver-enterprise-asm</artifactId>
+  </dependency>
+</dependencies>
+<!-- ... -->
+----
++
+* Add `<constraintStreamAutomaticNodeSharing>true</constraintStreamAutomaticNodeSharing>` in your `solverConfig.xml`:
++
+[source,xml,options="nowrap"]
+----
+<!-- ... -->
+<scoreDirectorFactory>
+  <constraintProviderClass>org.acme.MyConstraintProvider</constraintProviderClass>
+  <constraintStreamAutomaticNodeSharing>true</constraintStreamAutomaticNodeSharing>
+</scoreDirectorFactory>
+<!-- ... -->
+----
+
+Quarkus::
+Automatic node sharing is enabled by default.
```

Comment:
If we're not enabling it by default everywhere, then I'd argue it shouldn't be the default anywhere.